File: packages/types/src/calendar.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

export interface ICalendarRange {
  startDate: Date;
  endDate: Date;
}

export interface ICalendarDate {
  date: Date;
  year: number;
  month: number;
  day: number;
  week: number; // week number wrt year, eg- 51, 52
  is_current_month: boolean;
  is_current_week: boolean;
  is_today: boolean;
}

export interface ICalendarWeek {
  [date: string]: ICalendarDate;
}

export interface ICalendarMonth {
  [monthIndex: string]: {
    [weekNumber: string]: ICalendarWeek;
  };
}

export interface ICalendarPayload {
  [year: string]: ICalendarMonth;
}

export type TMeetingVisibility = "PROJECT" | "RESTRICTED" | "PERSONAL";
export type TMeetingStatus = "PLANNED" | "COMPLETED" | "CANCELLED";
export type TMeetingAttendanceMode = "ONLINE" | "OFFLINE" | "MIXED";
export type TMeetingResponseStatus = "NO_RESPONSE" | "ACCEPTED" | "TENTATIVE" | "DECLINED";
export type TMeetingParticipantRole = "REQUIRED" | "OPTIONAL";
export type TCalendarView = "DAY" | "WEEK" | "MONTH" | "SCHEDULE";

export interface TCalendarPreference {
  id: string;
  default_view: TCalendarView;
  filters: Record<string, unknown>;
  external_routing: Record<string, unknown>;
  default_reminders: number[];
  working_hours: Record<string, { enabled: boolean; start: string; end: string }>;
  show_weekends: boolean;
  email_notifications_enabled: boolean;
  in_app_notifications_enabled: boolean;
  updated_at: string;
}

export interface TCalendarUser {
  id: string;
  first_name: string;
  last_name: string;
  email: string;
  avatar_url?: string | null;
  user_timezone: string;
}

export interface TMeetingParticipant {
  id?: string;
  user?: TCalendarUser | null;
  user_id?: string;
  email: string;
  name: string;
  role: TMeetingParticipantRole;
  response_status?: TMeetingResponseStatus;
  source?: string;
  source_identifier?: string | null;
  responded_at?: string | null;
  removed_at?: string | null;
}

export interface TMeetingType {
  id: string;
  name: string;
  color: string;
  icon: string;
  default_duration_minutes: number;
  default_reminders: number[];
  default_attendance_mode: TMeetingAttendanceMode;
  is_default: boolean;
  sort_order: number;
}

export interface TMeeting {
  id: string;
  workspace_id?: string | null;
  project_id?: string | null;
  project_identifier?: string | null;
  project_name?: string | null;
  issue_id?: string | null;
  issue_sequence_id?: number | null;
  meeting_type_id?: string | null;
  organizer: TCalendarUser;
  title: string;
  description?: string;
  agenda?: string;
  starts_at: string;
  ends_at: string;
  all_day: boolean;
  timezone: string;
  location?: string;
  attendance_mode: TMeetingAttendanceMode;
  meeting_url?: string;
  meeting_url_source: "NONE" | "MANUAL" | "GOOGLE_MEET";
  google_meet_open_access: boolean;
  visibility: TMeetingVisibility;
  availability: "BUSY" | "FREE" | "MAYBE" | "AWAY";
  status: TMeetingStatus;
  recurrence_rule?: string;
  recurrence_timezone?: string;
  recurrence_until?: string | null;
  type_name_snapshot?: string;
  type_color_snapshot?: string;
  type_icon_snapshot?: string;
  task_snapshot?: Record<string, unknown>;
  participant_details?: TMeetingParticipant[];
  reminders?: Array<{ id: string; minutes_before: number; scheduled_for: string; channel: string }>;
  comments?: Array<{ id: string; user: TCalendarUser; body: string; created_at: string }>;
  attachments?: TMeetingAttachment[];
  detail_access: boolean;
  can_manage: boolean;
  can_respond: boolean;
  my_response_status?: TMeetingResponseStatus | null;
  cancelled_at?: string | null;
  completed_at?: string | null;
}

export interface TMeetingAttachment {
  id: string;
  asset_id: string;
  name: string;
  size: number;
  shared_with_guests: boolean;
  url: string;
  can_delete: boolean;
  created_at: string;
}

export interface TCalendarAvailabilityResponse {
  busy: Record<string, Array<{ starts_at: string; ends_at: string; source: string }>>;
  suggested_slots: Array<{
    starts_at: string;
    ends_at: string;
    optional_conflicts: string[];
    outside_working_hours: string[];
    score: number;
  }>;
}

export interface TMeetingOccurrence {
  meeting_id: string;
  occurrence_id: string;
  original_starts_at: string;
  starts_at: string;
  ends_at: string;
}

export interface TWorkspaceHoliday {
  id: string;
  date: string;
  name: string;
  kind: "HOLIDAY" | "DAY_OFF" | "WORKDAY";
  source: "MY_GOV" | "MANUAL" | "CORPORATE";
  is_override: boolean;
}

export interface TCalendarExternalEvent {
  id: string;
  meeting_id?: string | null;
  provider: "GOOGLE" | "MICROSOFT" | "ICLOUD" | "CALDAV";
  title: string;
  starts_at: string;
  ends_at: string;
  all_day: boolean;
  availability: "BUSY" | "FREE" | "MAYBE" | "AWAY";
}

export interface TCalendarBirthdayEvent {
  id: string;
  date: string;
  user: {
    id: string;
    first_name: string;
    last_name: string;
    avatar_url?: string | null;
  };
}

export interface TCalendarRangeResponse {
  meetings: TMeeting[];
  occurrences: TMeetingOccurrence[];
  external_events: TCalendarExternalEvent[];
  holidays: TWorkspaceHoliday[];
  birthday_events: TCalendarBirthdayEvent[];
}

export interface TMeetingPayload {
  project_id?: string | null;
  issue_id?: string | null;
  meeting_type_id?: string | null;
  title: string;
  description?: string;
  agenda?: string;
  starts_at: string;
  ends_at: string;
  all_day?: boolean;
  timezone: string;
  location?: string;
  attendance_mode: TMeetingAttendanceMode;
  meeting_url?: string;
  google_meet_open_access?: boolean;
  visibility?: TMeetingVisibility;
  availability?: "BUSY" | "FREE" | "MAYBE" | "AWAY";
  recurrence_rule?: string;
  recurrence_timezone?: string;
  recurrence_until?: string | null;
  participants?: TMeetingParticipant[];
  reminder_minutes?: number[];
}

export interface TCalendarWorkspaceSettings {
  id: string;
  country_code: string;
  working_hours: Record<string, { enabled: boolean; start: string; end: string }>;
  holiday_sync_enabled: boolean;
  holiday_source_url: string;
  holiday_last_synced_at?: string | null;
  holiday_last_error?: string;
}

export interface TCalendarConnection {
  id: string;
  provider: "GOOGLE" | "MICROSOFT" | "ICLOUD" | "CALDAV";
  account_email: string;
  account_label: string;
  status: "CONNECTED" | "PARTIAL" | "PAUSED" | "ERROR";
  selected_calendars: string[];
  sync_mode: "FULL" | "INBOUND_BUSY" | "OUTBOUND_GTS" | "DISABLED";
  last_synced_at?: string | null;
  last_error_at?: string | null;
  last_error?: string;
  has_credentials: boolean;
}

File: apps/api/plane/db/models/calendar.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only

from django.conf import settings
from django.db import models
from django.db.models import Q
from django.utils import timezone

from .base import BaseModel


def get_default_working_hours():
    """Default Monday-Friday schedule in the user's own timezone."""
    return {str(day): {"enabled": day <= 5, "start": "08:00", "end": "22:00"} for day in range(1, 8)}


def get_default_calendar_filters():
    return {
        "project_id": "",
        "organizer_id": "",
        "show_cancelled": False,
        "show_external": True,
    }


class MeetingVisibility(models.TextChoices):
    PROJECT = "PROJECT", "Project"
    RESTRICTED = "RESTRICTED", "Restricted"
    PERSONAL = "PERSONAL", "Personal"


class MeetingStatus(models.TextChoices):
    PLANNED = "PLANNED", "Planned"
    COMPLETED = "COMPLETED", "Completed"
    CANCELLED = "CANCELLED", "Cancelled"


class MeetingAttendanceMode(models.TextChoices):
    ONLINE = "ONLINE", "Online"
    OFFLINE = "OFFLINE", "Offline"
    MIXED = "MIXED", "Mixed"


class MeetingAvailability(models.TextChoices):
    BUSY = "BUSY", "Busy"
    FREE = "FREE", "Free"
    MAYBE = "MAYBE", "Maybe"
    AWAY = "AWAY", "Away"


class MeetingLinkSource(models.TextChoices):
    NONE = "NONE", "None"
    MANUAL = "MANUAL", "Manual"
    GOOGLE_MEET = "GOOGLE_MEET", "Google Meet"


class MeetingParticipantRole(models.TextChoices):
    REQUIRED = "REQUIRED", "Required"
    OPTIONAL = "OPTIONAL", "Optional"


class MeetingResponseStatus(models.TextChoices):
    NO_RESPONSE = "NO_RESPONSE", "No response"
    ACCEPTED = "ACCEPTED", "Accepted"
    TENTATIVE = "TENTATIVE", "Tentative"
    DECLINED = "DECLINED", "Declined"


class MeetingParticipantSource(models.TextChoices):
    ORGANIZER = "ORGANIZER", "Organizer"
    EXPLICIT = "EXPLICIT", "Explicit"
    TASK_CREATOR = "TASK_CREATOR", "Task creator"
    TASK_ASSIGNEE = "TASK_ASSIGNEE", "Task assignee"
    TASK_PROPERTY = "TASK_PROPERTY", "Task member property"
    TASK_SUBSCRIBER = "TASK_SUBSCRIBER", "Task subscriber"
    TASK_GROUP = "TASK_GROUP", "Task access group"


class CalendarProvider(models.TextChoices):
    GOOGLE = "GOOGLE", "Google Calendar"
    MICROSOFT = "MICROSOFT", "Microsoft Outlook"
    ICLOUD = "ICLOUD", "iCloud"
    CALDAV = "CALDAV", "CalDAV"


class CalendarSyncMode(models.TextChoices):
    FULL = "FULL", "Bidirectional"
    INBOUND_BUSY = "INBOUND_BUSY", "Inbound busy only"
    OUTBOUND_GTS = "OUTBOUND_GTS", "GTS outbound only"
    DISABLED = "DISABLED", "Disabled"


class CalendarConnectionStatus(models.TextChoices):
    CONNECTED = "CONNECTED", "Connected"
    PARTIAL = "PARTIAL", "Partially synchronized"
    PAUSED = "PAUSED", "Paused"
    ERROR = "ERROR", "Error"


class WorkspaceCalendarSettings(BaseModel):
    workspace = models.OneToOneField(
        "db.Workspace",
        on_delete=models.CASCADE,
        related_name="calendar_settings",
    )
    country_code = models.CharField(max_length=2, default="UZ")
    working_hours = models.JSONField(default=get_default_working_hours)
    holiday_sync_enabled = models.BooleanField(default=True)
    holiday_source_url = models.URLField(max_length=800, default="https://my.gov.uz/ru/day-off")
    holiday_last_synced_at = models.DateTimeField(null=True, blank=True)
    holiday_last_error = models.TextField(blank=True, default="")

    class Meta:
        db_table = "workspace_calendar_settings"


class WorkspaceHoliday(BaseModel):
    class Kind(models.TextChoices):
        HOLIDAY = "HOLIDAY", "Official holiday"
        DAY_OFF = "DAY_OFF", "Day off"
        WORKDAY = "WORKDAY", "Working day"

    class Source(models.TextChoices):
        MY_GOV = "MY_GOV", "my.gov.uz"
        MANUAL = "MANUAL", "Manual override"
        CORPORATE = "CORPORATE", "Corporate"

    workspace = models.ForeignKey(
        "db.Workspace",
        on_delete=models.CASCADE,
        related_name="calendar_holidays",
    )
    date = models.DateField()
    name = models.CharField(max_length=255, blank=True, default="")
    kind = models.CharField(max_length=16, choices=Kind.choices)
    source = models.CharField(max_length=16, choices=Source.choices, default=Source.MY_GOV)
    source_key = models.CharField(max_length=255, blank=True, default="")
    source_payload = models.JSONField(default=dict, blank=True)
    is_override = models.BooleanField(default=False)

    class Meta:
        db_table = "workspace_holidays"
        ordering = ("date",)
        constraints = [
            models.UniqueConstraint(
                fields=["workspace", "date", "source"],
                condition=Q(deleted_at__isnull=True),
                name="unique_active_workspace_holiday_source",
            )
        ]
        indexes = [models.Index(fields=["workspace", "date"], name="workspace_holiday_date_idx")]


class BirthdayNotificationDelivery(BaseModel):
    class DeliveryType(models.TextChoices):
        ADVANCE_IN_APP = "ADVANCE_IN_APP", "Advance in-app notification"
        ADVANCE_EMAIL = "ADVANCE_EMAIL", "Advance email notification"
        ADVANCE_TELEGRAM = "ADVANCE_TELEGRAM", "Advance Telegram notification"
        SELF_EMAIL = "SELF_EMAIL", "Birthday email greeting"
        SELF_TELEGRAM = "SELF_TELEGRAM", "Birthday Telegram greeting"

    recipient = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.CASCADE,
        related_name="birthday_notifications_received",
    )
    birthday_user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.CASCADE,
        related_name="birthday_notifications_about",
    )
    occurrence_year = models.PositiveSmallIntegerField()
    delivery_type = models.CharField(max_length=32, choices=DeliveryType.choices)
    delivered_at = models.DateTimeField(default=timezone.now)

    class Meta:
        db_table = "birthday_notification_deliveries"
        constraints = [
            models.UniqueConstraint(
                fields=["recipient", "birthday_user", "occurrence_year", "delivery_type"],
                condition=Q(deleted_at__isnull=True),
                name="unique_active_birthday_delivery",
            )
        ]
        indexes = [
            models.Index(
                fields=["birthday_user", "occurrence_year"],
                name="bday_delivery_occurrence_idx",
            )
        ]


class MeetingType(BaseModel):
    workspace = models.ForeignKey(
        "db.Workspace",
        on_delete=models.CASCADE,
        related_name="meeting_types",
    )
    project = models.ForeignKey(
        "db.Project",
        on_delete=models.CASCADE,
        related_name="meeting_types",
    )
    name = models.CharField(max_length=120)
    color = models.CharField(max_length=16, default="#22C55E")
    icon = models.CharField(max_length=64, default="calendar")
    default_duration_minutes = models.PositiveIntegerField(default=30)
    default_reminders = models.JSONField(default=list, blank=True)
    default_attendance_mode = models.CharField(
        max_length=16,
        choices=MeetingAttendanceMode.choices,
        default=MeetingAttendanceMode.ONLINE,
    )
    is_default = models.BooleanField(default=False)
    sort_order = models.FloatField(default=65535)

    class Meta:
        db_table = "meeting_types"
        ordering = ("sort_order", "name")
        constraints = [
            models.UniqueConstraint(
                fields=["project", "name"],
                condition=Q(deleted_at__isnull=True),
                name="unique_active_project_meeting_type_name",
            ),
            models.UniqueConstraint(
                fields=["project"],
                condition=Q(is_default=True, deleted_at__isnull=True),
                name="unique_default_project_meeting_type",
            ),
        ]


class CalendarPreference(BaseModel):
    user = models.OneToOneField(
        settings.AUTH_USER_MODEL,
        on_delete=models.CASCADE,
        related_name="calendar_preference",
    )
    default_view = models.CharField(max_length=16, default="MONTH")
    filters = models.JSONField(default=get_default_calendar_filters)
    external_routing = models.JSONField(default=dict, blank=True)
    default_reminders = models.JSONField(default=list, blank=True)
    working_hours = models.JSONField(default=get_default_working_hours)
    show_weekends = models.BooleanField(default=True)
    email_notifications_enabled = models.BooleanField(default=True)
    in_app_notifications_enabled = models.BooleanField(default=True)

    class Meta:
        db_table = "calendar_preferences"


class Meeting(BaseModel):
    workspace = models.ForeignKey(
        "db.Workspace",
        on_delete=models.SET_NULL,
        related_name="meetings",
        null=True,
        blank=True,
    )
    project = models.ForeignKey(
        "db.Project",
        on_delete=models.SET_NULL,
        related_name="meetings",
        null=True,
        blank=True,
    )
    issue = models.ForeignKey(
        "db.Issue",
        on_delete=models.SET_NULL,
        related_name="meetings",
        null=True,
        blank=True,
    )
    meeting_type = models.ForeignKey(
        MeetingType,
        on_delete=models.SET_NULL,
        related_name="meetings",
        null=True,
        blank=True,
    )
    organizer = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.PROTECT,
        related_name="organized_meetings",
    )
    title = models.CharField(max_length=255)
    description = models.TextField(blank=True, default="")
    agenda = models.TextField(blank=True, default="")
    starts_at = models.DateTimeField(db_index=True)
    ends_at = models.DateTimeField(db_index=True)
    all_day = models.BooleanField(default=False)
    timezone = models.CharField(max_length=255, default="UTC")
    location = models.CharField(max_length=500, blank=True, default="")
    attendance_mode = models.CharField(
        max_length=16,
        choices=MeetingAttendanceMode.choices,
        default=MeetingAttendanceMode.ONLINE,
    )
    meeting_url = models.URLField(max_length=1000, blank=True, default="")
    meeting_url_source = models.CharField(
        max_length=16,
        choices=MeetingLinkSource.choices,
        default=MeetingLinkSource.NONE,
    )
    google_meet_space = models.CharField(max_length=255, blank=True, default="")
    google_meet_open_access = models.BooleanField(default=True)
    visibility = models.CharField(
        max_length=16,
        choices=MeetingVisibility.choices,
        default=MeetingVisibility.PROJECT,
        db_index=True,
    )
    availability = models.CharField(
        max_length=16,
        choices=MeetingAvailability.choices,
        default=MeetingAvailability.BUSY,
    )
    status = models.CharField(
        max_length=16,
        choices=MeetingStatus.choices,
        default=MeetingStatus.PLANNED,
        db_index=True,
    )
    recurrence_rule = models.TextField(blank=True, default="")
    recurrence_timezone = models.CharField(max_length=255, blank=True, default="")
    recurrence_until = models.DateTimeField(null=True, blank=True)
    type_name_snapshot = models.CharField(max_length=120, blank=True, default="")
    type_color_snapshot = models.CharField(max_length=16, blank=True, default="")
    type_icon_snapshot = models.CharField(max_length=64, blank=True, default="")
    cancelled_at = models.DateTimeField(null=True, blank=True)
    completed_at = models.DateTimeField(null=True, blank=True)
    task_snapshot = models.JSONField(default=dict, blank=True)
    public_token_hash = models.CharField(max_length=64, blank=True, default="", db_index=True)

    class Meta:
        db_table = "meetings"
        ordering = ("starts_at",)
        indexes = [
            models.Index(fields=["organizer", "starts_at"], name="meeting_organizer_start_idx"),
            models.Index(fields=["workspace", "starts_at", "ends_at"], name="meeting_workspace_range_idx"),
            models.Index(fields=["project", "starts_at", "ends_at"], name="meeting_project_range_idx"),
            models.Index(fields=["issue", "starts_at"], name="meeting_issue_start_idx"),
        ]

    def save(self, *args, **kwargs):
        if self.project_id:
            self.workspace_id = self.project.workspace_id
        if self.meeting_type_id and not self.type_name_snapshot:
            self.type_name_snapshot = self.meeting_type.name
            self.type_color_snapshot = self.meeting_type.color
            self.type_icon_snapshot = self.meeting_type.icon
        super().save(*args, **kwargs)


class MeetingParticipant(BaseModel):
    meeting = models.ForeignKey(Meeting, on_delete=models.CASCADE, related_name="participants")
    user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.SET_NULL,
        related_name="meeting_participations",
        null=True,
        blank=True,
    )
    email = models.EmailField(max_length=255)
    name = models.CharField(max_length=255, blank=True, default="")
    role = models.CharField(
        max_length=16,
        choices=MeetingParticipantRole.choices,
        default=MeetingParticipantRole.REQUIRED,
    )
    response_status = models.CharField(
        max_length=16,
        choices=MeetingResponseStatus.choices,
        default=MeetingResponseStatus.NO_RESPONSE,
    )
    source = models.CharField(
        max_length=24,
        choices=MeetingParticipantSource.choices,
        default=MeetingParticipantSource.EXPLICIT,
    )
    source_identifier = models.UUIDField(null=True, blank=True)
    response_token_hash = models.CharField(max_length=64, unique=True, db_index=True)
    responded_at = models.DateTimeField(null=True, blank=True)
    removed_at = models.DateTimeField(null=True, blank=True)
    notify_by_email = models.BooleanField(default=True)

    class Meta:
        db_table = "meeting_participants"
        ordering = ("created_at",)
        constraints = [
            models.UniqueConstraint(
                fields=["meeting", "user"],
                condition=Q(user__isnull=False, deleted_at__isnull=True),
                name="unique_active_meeting_user_participant",
            ),
            models.UniqueConstraint(
                fields=["meeting", "email"],
                condition=Q(deleted_at__isnull=True),
                name="unique_active_meeting_email_participant",
            ),
        ]
        indexes = [
            models.Index(fields=["user", "response_status"], name="meeting_participant_user_idx"),
            models.Index(fields=["meeting", "removed_at"], name="meeting_participant_active_idx"),
        ]


class MeetingReminder(BaseModel):
    meeting = models.ForeignKey(Meeting, on_delete=models.CASCADE, related_name="reminders")
    participant = models.ForeignKey(
        MeetingParticipant,
        on_delete=models.CASCADE,
        related_name="reminders",
        null=True,
        blank=True,
    )
    minutes_before = models.PositiveIntegerField()
    scheduled_for = models.DateTimeField(db_index=True)
    sent_at = models.DateTimeField(null=True, blank=True)
    channel = models.CharField(max_length=16, default="IN_APP")

    class Meta:
        db_table = "meeting_reminders"
        indexes = [models.Index(fields=["sent_at", "scheduled_for"], name="meeting_reminder_due_idx")]


class MeetingRecurrenceException(BaseModel):
    class Action(models.TextChoices):
        UPDATED = "UPDATED", "Updated"
        CANCELLED = "CANCELLED", "Cancelled"

    meeting = models.ForeignKey(Meeting, on_delete=models.CASCADE, related_name="recurrence_exceptions")
    original_starts_at = models.DateTimeField()
    action = models.CharField(max_length=16, choices=Action.choices)
    overrides = models.JSONField(default=dict, blank=True)

    class Meta:
        db_table = "meeting_recurrence_exceptions"
        constraints = [
            models.UniqueConstraint(
                fields=["meeting", "original_starts_at"],
                condition=Q(deleted_at__isnull=True),
                name="unique_active_meeting_recurrence_exception",
            )
        ]


class MeetingComment(BaseModel):
    meeting = models.ForeignKey(Meeting, on_delete=models.CASCADE, related_name="comments")
    user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.SET_NULL,
        related_name="meeting_comments",
        null=True,
    )
    body = models.TextField()

    class Meta:
        db_table = "meeting_comments"
        ordering = ("created_at",)


class MeetingActivity(BaseModel):
    meeting = models.ForeignKey(Meeting, on_delete=models.CASCADE, related_name="activities")
    actor = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.SET_NULL,
        related_name="meeting_activities",
        null=True,
    )
    event = models.CharField(max_length=64)
    details = models.JSONField(default=dict, blank=True)

    class Meta:
        db_table = "meeting_activities"
        ordering = ("-created_at",)
        indexes = [models.Index(fields=["meeting", "created_at"], name="meeting_activity_time_idx")]


class MeetingAttachment(BaseModel):
    meeting = models.ForeignKey(Meeting, on_delete=models.CASCADE, related_name="attachments")
    asset = models.OneToOneField(
        "db.FileAsset",
        on_delete=models.CASCADE,
        related_name="meeting_attachment",
    )
    shared_with_guests = models.BooleanField(default=False)

    class Meta:
        db_table = "meeting_attachments"


class CalendarConnection(BaseModel):
    user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.CASCADE,
        related_name="calendar_connections",
    )
    provider = models.CharField(max_length=16, choices=CalendarProvider.choices)
    account_email = models.EmailField(max_length=255, blank=True, default="")
    account_label = models.CharField(max_length=255, blank=True, default="")
    status = models.CharField(
        max_length=16,
        choices=CalendarConnectionStatus.choices,
        default=CalendarConnectionStatus.CONNECTED,
    )
    credentials_encrypted = models.TextField(blank=True, default="")
    server_url_encrypted = models.TextField(blank=True, default="")
    selected_calendars = models.JSONField(default=list, blank=True)
    sync_mode = models.CharField(
        max_length=24,
        choices=CalendarSyncMode.choices,
        default=CalendarSyncMode.FULL,
    )
    sync_token_encrypted = models.TextField(blank=True, default="")
    webhook_channel_id = models.CharField(max_length=255, blank=True, default="")
    webhook_expires_at = models.DateTimeField(null=True, blank=True)
    last_synced_at = models.DateTimeField(null=True, blank=True)
    last_error_at = models.DateTimeField(null=True, blank=True)
    last_error = models.TextField(blank=True, default="")

    class Meta:
        db_table = "calendar_connections"
        ordering = ("provider", "account_email")
        constraints = [
            models.UniqueConstraint(
                fields=["user", "provider", "account_email"],
                condition=Q(deleted_at__isnull=True),
                name="unique_active_user_calendar_connection",
            )
        ]


class MeetingExternalEvent(BaseModel):
    meeting = models.ForeignKey(
        Meeting,
        on_delete=models.CASCADE,
        related_name="external_events",
        null=True,
        blank=True,
    )
    connection = models.ForeignKey(
        CalendarConnection,
        on_delete=models.CASCADE,
        related_name="external_events",
    )
    provider_event_id = models.CharField(max_length=512)
    provider_calendar_id = models.CharField(max_length=512, blank=True, default="")
    title = models.CharField(max_length=255, blank=True, default="")
    starts_at = models.DateTimeField(db_index=True)
    ends_at = models.DateTimeField(db_index=True)
    all_day = models.BooleanField(default=False)
    availability = models.CharField(
        max_length=16,
        choices=MeetingAvailability.choices,
        default=MeetingAvailability.BUSY,
    )
    provider_updated_at = models.DateTimeField(null=True, blank=True)
    etag = models.CharField(max_length=512, blank=True, default="")
    raw_payload_encrypted = models.TextField(blank=True, default="")
    is_deleted_at_provider = models.BooleanField(default=False)

    class Meta:
        db_table = "meeting_external_events"
        constraints = [
            models.UniqueConstraint(
                fields=["connection", "provider_calendar_id", "provider_event_id"],
                condition=Q(deleted_at__isnull=True),
                name="unique_active_calendar_external_event",
            )
        ]
        indexes = [models.Index(fields=["connection", "starts_at", "ends_at"], name="external_event_range_idx")]
